server/src/plugins/google-storage-utils.ts

```diff
@@ -0,0 +1,114 @@
+
+import * as storage from '@google-cloud/storage';
+import getEnv from '../envs';
+
+function sleep(ms: number) {
+  return new Promise(resolve => setTimeout(resolve, ms));
+}
+
+export async function uploadFile(gcsFileName: string, mimetype:string, in_memory_file:any): Promise<string | null> {
+    const env = getEnv();
+    let finished = false;
+    const googleStorageClient = new storage.Storage();
+    const bucketName = env.googleStorageBucket;
+    const bucket = googleStorageClient.bucket(bucketName);
+    const file = bucket.file(gcsFileName);
+  
+    const blob = file.createWriteStream({
+      metadata: {
+        contentType: mimetype,
+      },
+      public: true
+    });
+
+    blob.on('error', (err: any) => {
+      console.log(err);
+      return null;
+    });
+  
+    blob.on('finish', () => {
+      finished = true;
+    });
+
+    blob.end(in_memory_file);
+
+    while(finished===false){
+      await sleep(200);
+    }
+
+    return `https://storage.googleapis.com/${bucket.name}/${gcsFileName}`
```

Comment:
My best guess for this code is to "promisify" the blob object so we can use `async/await` later and avoid the use of `while (finished === false) { await sleep(200); }`

Code example:
```typescript
const blob = file.createWriteStream({ metadata: { contentType: mimetype }, public: true });

// This helper can be moved outside the uploadFile function if you desire
const promisifyBlob = (newBlob, file) => {
  return new Promise((resolve, reject) => {
    newBlob.on('error', error => reject(error));
    newBlob.on('finish', () => resolve(blob.end(file)));
  });
};

// This will finish when the 'finish' event is triggered:
const promisifiedBlob = await promisifyBlob(blob, in_memory_file);

return `https://storage.googleapis.com/${bucket.name}/${gcsFileName}`
```

Reply:
`blob.end(in_memory_file)` triggers `blob.on('finish', () => {` i'm wrong or if you put `blob.end` inside `finish` the `end` will never execute?

Reply:
Get it! Let me recap on this one. 

1. Let's send to the helper the file and the inMemory file.
2. Would be better if we create the stream inside the promise.
3. Trigger the `end(inMemoryFile)` inside the promise.
4. Trigger the `resolve()` method on `finish` event.

```typescript
// This helper can be moved outside the uploadFile function if you desire
const promisifyBlob = (file: any, inMemoryFile: any) => {
  return new Promise((resolve, reject) => {
    const blob = file.createWriteStream({ metadata: { contentType: mimetype }, public: true });
    blob.end(inMemoryFile);

    blob.on('error', error => reject(error));
    blob.on('finish', () => resolve());
  });
};

// This will finish when the 'finish' event is triggered:
await promisifyBlob(file, in_memory_file);

return `https://storage.googleapis.com/${bucket.name}/${gcsFileName}`
}
```

It's not tested but let me know if find some bugs or improvements!